test(cypress): add cypress tests and move getMetaTagValue to utils

luqven opened the pull request:
<!--
Hello, and thanks for contributing to imgix.js! 🎉🙌
Please take a second to fill out PRs with the following template!
-->

# Description

This PR adds [cypress](cypress.io) integration testing capabilities. It also adds a cypress [test](https://github.com/imgix/imgix.js/blob/luis/metaTagRefactor/cypress/integration/metaSpec.js) for [`getMetaTagValue()`](https://github.com/imgix/imgix.js/blob/76ffa7ee6158c13290260439966590c7ae6d172c/src/util.js#L83-L105). Finally, the PR moves the `getMetaTagValue` function from the `imgix.js` file into the `util.js` file.

<!-- What is accomplished by this PR? If there is something potentially controversial in your PR, please take a moment to tell us about your choices. -->

The goal of this PR is to enable future integration tests in this package with cypress. The decision to move `getMetaTagValue` function could prove controversial since it's outside of initial scope for the PR. However, that was done to remove clutter from `imgix.js` and to follow what appears to be the standard of having this DOM related helper functions in the `util.js` file.

<!-- 
Please use the checklist that is most closely related to your PR, and delete the other checklists. -->

## Checklist: Fixing typos/Doc change

- [x] Each commit follows the [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/#summary) format: e.g. `chore(readme): fixed typo`. See the end of this file for more information.

## Checklist: New Feature

- [x] Each commit follows the [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/#summary) format: e.g. `chore(readme): fixed typo`. See the end of this file for more information.
- [x] Any breaking changes are specified on the commit on which they are introduced with `BREAKING CHANGE` in the body of the commit.
- [x] If this is a big feature with breaking changes, consider [opening an issue](https://github.com/imgix/imgix.js/issues/new?assignees=&labels=&template=feature_request.md&title=) to discuss first. This is completely up to you, but please keep in mind that your PR might not be accepted.
- [x] Run unit tests to ensure all existing tests are still passing
- [x] Add new passing unit tests to cover the code introduced by your PR
- [ ] Update the readme
- [x] Add some [steps](#steps-to-test) so we can test your cool new feature!

## Steps to Test

<!-- Delete this selction if you are just submitting a doc change/small fix -->
- `npm run cy:run`
<!-- A code example or a set of steps is preferred -->
<!-- A link to a codepen/codesandbox is also an option. -->

<!--

## Conventional Commit Spec

PR titles should be in the format `<type>(<scope>): <description>`. For example: `chore(readme): fix typo`

`type` can be any of the follow:
  - `feat`: a feature, or breaking change
  - `fix`: a bug-fix
  - `test`: Adding missing tests or correcting existing tests
  - `docs`: documentation only changes (readme, changelog, contributing guide)
  - `refactor`: a code change that neither fixes a bug nor adds a feature
  - `chore`: reoccurring tasks for project maintainability (example scopes: release, deps)
  - `config`: changes to tooling configurations used in the project
  - `build`: changes that affect the build system or external dependencies (example scopes: npm, bundler, gradle)
  - `ci`: changes to CI configuration files and scripts (example scopes: travis)
  - `perf`: a code change that improves performance
  - `style`: changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)

`scope` is optional, and can be anything.
`description` should be a short description of the change, written in the imperative-mood.
-->

sherwinski commented on cypress/integration/metaSpec.js:
```diff
@@ -0,0 +1,27 @@
+    });
+  });
+  it('successfully applies imgix paramenters to the source url', () => {
+    cy.get('img').each(($el, index, $list) => {
```
Nit: we should remove `$list` if it's not getting referenced in the callback fn
```suggestion
    cy.get('img').each(($el, index) => {
```

sherwinski reviewed (changes requested):
This looks great! 👍 

As a general note, we try not to check the `dist` folder into pull requests as it can cause a lot of messiness down the line. However, we do generate them during the library release step. Aside from that, I just have a couple of nitpicks which I'll go into more detail about below.

Also a separate question, do you think we should integrate `npm run cy:run` into our CI and/or extend `npm test` to run the cypress tests as well?

luqven commented:
> As a general note, we try not to check the `dist` folder into pull requests as it can cause a lot of messiness down the line. However, we do generate them during the library release step. Aside from that, I just have a couple of nitpicks which I'll go into more detail about below.

Thanks for the feedback, will address these comments asap.

Ah ms not the check that in, noted. 

> Also a separate question, do you think we should integrate `npm run cy:run` into our CI and/or extend `npm test` to run the cypress tests as well?
And I'm all for adding cypress to `npm run test` 👌🏼

frederickfogerty commented on cypress/integration/metaSpec.js:
```diff
@@ -0,0 +1,27 @@
+describe('The example page with imgix meta tags', () => {
+  it('successfully loads', () => {
```
I'm not really sure what this is testing - it might be better to include a more descriptive test name.

If I'm understanding the test correctly, it seems like it's just testing the behaviour of Cypress itself, which I don't think is necessary to test? If so this test might not be necessary?

luqven replied:
This is a good point, I meant to test that when the page has a meta tag imgix.js was aware of its existence, ie `util.getMetaTagValue(key)` does not return `undefined`. As written this test doesn't do that. Will fix.

frederickfogerty commented on cypress/integration/metaSpec.js:
```diff
@@ -0,0 +1,27 @@
+      }
+    });
+  });
+  it('successfully applies imgix paramenters to the source url', () => {
```
This test is brittle for a few reasons:

- if the rendered url changes the order of its query parameters then the test breaks
- if the package version changes the test breaks

I would recommend using something like the `URL` API and checking that the parameters are applied.

---

From a different perspective, I'm not sure how valuable this is as a test since we already have tests for parameters being set. An idea could be to simply test if the image loads, since this is then improving on our existing test suite. An example of a test like this can be seen in other Cypress tests we have, such as this one: https://github.com/imgix/gatsby/blob/main/test/dev-and-e2e/cypress/e2e/gatsby-transform-node.js

frederickfogerty commented on src/util.js:
```diff
@@ -60,24 +60,46 @@ module.exports = {
     }
-  }
+  },
+  getMetaTagValue: function (propertyName) {
```
I like this refactor, it makes a lot more sense to have this utility located here, nice work 👏

frederickfogerty commented on cypress/integration/metaSpec.js:
```diff
@@ -0,0 +1,27 @@
+describe('The example page with imgix meta tags', () => {
+  it('successfully loads', () => {
+    cy.visit('/cypress/fixtures/samplePage.html');
+    // .should('equal', 'ix:useHttps');
```
```suggestion
```

frederickfogerty commented on cypress/integration/metaSpec.js:
```diff
@@ -0,0 +1,27 @@
+    });
+  });
+  it('successfully applies imgix paramenters to the source url', () => {
+    cy.get('img').each(($el, index) => {
```
If there are no `img`s on the page then this test succeeds when it should fail. This can be improved on by checking that the number of `img`s that are queried are the same as expected. The test above also suffers from a similar isssue if there are no meta tags.

frederickfogerty reviewed (changes requested):
This is a great addition to the library and I'm excited to get this in! I think we have a few things that we need to resolve before that can happen.

frederickfogerty commented on cypress/integration/metaSpec.js:
```diff
@@ -0,0 +1,51 @@
+    });
+    it('All loaded and rendered', () => {
+      cy.get('img').each(($el, idx, $list) => {
+        // epect the number of image elements on the page to equal fixture's
```
```suggestion
        // expect the number of image elements on the page to equal fixture's
```

frederickfogerty commented on cypress/integration/metaSpec.js:
```diff
@@ -0,0 +1,51 @@
+    it('All loaded and rendered', () => {
+      cy.get('img').each(($el, idx, $list) => {
+        // epect the number of image elements on the page to equal fixture's
+        idx === 0 && expect($list.length).to.equal(3);
```
I think this is better outside the loop as a separate assertion since this won't be checked if there are no images.

luqven replied:
yup yup, hair brain check there on my part 🤣  Just gonna throw up a `have.length` check above this loop 👍🏼

frederickfogerty commented on cypress/integration/metaSpec.js:
```diff
@@ -0,0 +1,51 @@
+    it('Interacted with the imgix instance', () => {
+      cy.get('img').each(($el) => {
+        expect($el.attr('ix-initialized')).to.equal('ix-initialized');
+      });
```
Same as above, we need to have a length check here so this won't silently fail

frederickfogerty reviewed (changes requested):
Looking better 👍

frederickfogerty commented on cypress/integration/metaSpec.js:
```diff
@@ -8,15 +8,16 @@ describe('On a page with meta tag imgix paramaters', () => {
     });
 
     it('Interacted with the imgix instance', () => {
+      cy.get('img').should('have.length', 3);
```
You probably want to have a similar check above line 38 for that test 👍

luqven replied:
Yup, just going to move this check to the beforeEach hook so that it runs on every test 👌🏼

frederickfogerty reviewed (approved):
Looking good!

frederickfogerty commented:
Also, similar to to the other PR, this PR's title and commits should use the proper commit message type - they should use either `test` or `chore`. `fix` and `feat` are reserved for changes that affect end users

luqven commented:
> Also, similar to to the other PR, this PR's title and commits should use the proper commit message type - they should use either `test` or `chore`. `fix` and `feat` are reserved for changes that affect end users

Got it. Rebased and changed the titles to be chore or test. Unsure if moving the meta tag to utils was a chore, but since it's not user-facing really my gut says it was.

Thanks for looking over this.. again! 😬

luqven commented:
Ah! Didn't realize the branch was 1 commit behind master. Had to rebase 1 last time to pull that in. Should be rtg. Sorry all for so much back and forth.
